[tdpl] Multiple alias this

IgorStepanov opened the pull request:
This PR introduces multiple alias this.
The old alias this implemenation doesn't imply alias this conflicts.
All alias this-es in old implementation are serializable: compiler successively applies a, a->aliasthis, a->aliasthis->aliasthis while right expression is not found.

This implementation allows tree of alias this subtypes.
Type can has a direct "alias this"-es and "alias this"-es, inherited from base classes/interfaces. Thus, there can be conflict between several matching "alias this"-es.
In common case compile should raise error if several "alias this"-es have found.
However, in the next situation conflict can be resolved:

```
struct Test9a
{
    int a;
    alias a this;
}

struct Test9b
{
    int a;
    alias a this;
}

struct Test9
{
    Test9a a;
    Test9b b;
    int c;
    alias a this;
    alias b this;
    alias c this;
}

void test()
{
    Test9 t9;
    int a = t9;
}
```

When we try to cast `Test9` to int we can find several variants: `t9.a.a`, `t9.b.a`, `t9.c`.
However, `t9.c` is preferred, because `Test9` author is explicitly pointed out that `Test9` should be converted to `int` through `c` field. This magic works only if we have exact match. `short a = t9;` causes error.

This PR have a one limitation: it doesn't support multiple tuple alias this. It don't clearly understand how them should works.   Multiple tuple alias this can be implemented in future in a separate PR.

Commit:
implement multiple alias this

MartinNowak commented:
What's the state of this? Wasn't the discussion finished and DIP66 accepted?

acehreli commented:
There seems to be merge conflicts and a failing test.

Ali

quickfur commented:
Can somebody update this PR to fix the merge conflicts and failing tests? It would be a big pity for this to go to waste.

IgorStepanov commented:
Hey, I still work on this.
@yebblies Can you help me with this error message (with your d port)?
```
core.exception.AssertError@magicport/ast.d(779): ast.TemplateType
```

yebblies commented:
I'd guess the type matcher can't handle `Array<bool>` as a function parameter.  I'd suggest defining `Bools` in `arraytypes.h` and using that.

IgorStepanov commented:
@yebblies Now it write `Error: ./aliasthis.c(209): Unknown basic type Bools`.
I defined `Bools` in `arraytypes.h`, added it to `arraytype` module in `magicport.json`, added `arraytypes` import to `aliasthis` module in `magicport.json`. What I need to do now?

yebblies commented:
Add 'Bools' to the 'structTypes' list in the json file.

IgorStepanov commented:
@yebblies Ok. Now it writes:
```
core.exception.AssertError@magicport/magicport2.d(279): function atFindMemberScope*Expression*void*Expression**
----------------
0x80bb317 <magicport2.D[immutable(char)[]] magicport2.buildLongMap(ast.Module)+143> at /home/braddr/sandbox/at-client/pull-1569631-FreeBSD_32/dmd/src/magicport/magicport2
0x80ba634 <_Dmain+3844> at /home/braddr/sandbox/at-client/pull-1569631-FreeBSD_32/dmd/src/magicport/magicport2
0x80ed09a <_D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv+18> at /home/braddr/sandbox/at-client/pull-1569631-FreeBSD_32/dmd/src/magicport/magicport2
0x80ed010 <void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate())+24> at /home/braddr/sandbox/at-client/pull-1569631-FreeBSD_32/dmd/src/magicport/magicport2
0x80ed05f <void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll()+39> at /home/braddr/sandbox/at-client/pull-1569631-FreeBSD_32/dmd/src/magicport/magicport2
0x80ed010 <void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate())+24> at /home/braddr/sandbox/at-client/pull-1569631-FreeBSD_32/dmd/src/magicport/magicport2
0x80ecfa8 <_d_run_main+440> at /home/braddr/sandbox/at-client/pull-1569631-FreeBSD_32/dmd/src/magicport/magicport2
0x80c3e3c <main+20> at /home/braddr/sandbox/at-client/pull-1569631-FreeBSD_32/dmd/src/magicport/magicport2
0x80b96b7 <_start+167> at /home/braddr/sandbox/at-client/pull-1569631-FreeBSD_32/dmd/src/magicport/magicport2
0x80b9628 <_start+24> at /home/braddr/sandbox/at-client/pull-1569631-FreeBSD_32/dmd/src/magicport/magicport2
0x3 <???+0> at ???
```
What do I need to add to the json file?

yebblies commented:
You are declaring the function `atfindmember` twice, with the exact same arguments.  Easy solution is to change the name of one.

IgorStepanov commented:
```
unreferenced: function atFindOpUna
unreferenced: function atFindMember
unreferenced: function atToBoolean
unreferenced: function atFindDotIdCall
unreferenced: struct DoBothCtx
unreferenced: function atFindDotId
unreferenced: function getAliasThisTypes
duplicate: typedef Strings
unreferenced: function findAliasThisSubtypes
unreferenced: function atResolveForeach
unreferenced: function mergeAliasThis
unreferenced: function atFindOpUnaBin
unreferenced: function atFindOpUnaUna
unreferenced: struct FindMemberCtx
unreferenced: function atCastTo
unreferenced: function atFindOpExp2Bin
unreferenced: function resloveAliasThisForBinExp
unreferenced: function atBinExpRvl
unreferenced: function atFindSubtypes
unreferenced: function atCheckErrors
unreferenced: function atBinExpLvl
unreferenced: struct FindDotIdCtx
unreferenced: function iterateAliasThis
unreferenced: function atBinExpBoth
unreferenced: function aliasThisOf
unreferenced: function atFindMember2
```
Where should I declare those symbols? What should I do with "duplicate: typedef Strings"
The most of those symbols don't represented in header files and used only as `.c` internal.

yebblies commented:
Where you put it is good - the same place as in the C++ source.  Don't worry about the "duplicate: typedef Strings".  It's just a warning and allows avoiding a layer-violating import.

yebblies commented:
Is defining all of those similar functions really necessary?  It would be really nice if they could be collapsed a bit.

yebblies commented on src/expression.c:
```diff
@@ -52,6 +52,105 @@ Type *getTypeInfoType(Type *t, Scope *sc);
+bool atFindOpUnaBin(Scope *sc, Expression *e, void *ctx, Expression **outexpr)
+{
+    BinExp *e1 = (BinExp*)ctx;
+    e1 = (BinExp *)e1->copy();
```
You don't need an extra line for these, really.

IgorStepanov replied:
Hmm, yes. I scared of an additional pair of parentheseses :)

ibuclaw commented:
The function names are a bit unintuitive.

IgorStepanov commented:
Wow, is passes.

>Where you put it is good - the same place as in the C++ source. Don't worry about the "duplicate: typedef Strings". It's just a warning and allows avoiding a layer-violating import.

I understood, thank you. You magic works :)

>Is defining all of those similar functions really necessary? It would be really nice if they could be collapsed a bit.

I already collapsed this part where it was possible. Those functions follows the similar pattern (get alias this variation of `e`, combine it with some object, check is it ok and If it is ok, save the result), however they represent different situations.
Your commented function, for example, takes a bin expression and its first part, and tries to resolve alias this for the first term and substitute it to the original `BinExp`. It similar works for assign expression (`a = b;` => `a.aliasthis = b;`), or index expression (`a[b];` => `a.aliasthis[b];`), for example. However combining with `UnaExp` (including `CallExp`, `CastExp` et c.), need a bit another logic.
I have no idea, how reduce this part more.

>The function names are a bit unintuitive.

Yes, I see this trouble too. Now I want to adapt this PR to DIP66 (there are some rules, which is not implemented here yet) and brush code. After that, I hope, It will be much pretty a bit.

IgorStepanov commented:
I've finished preparing of the this PR. Now it is ready for review.

IgorStepanov commented:
@andralex Hey, time to review?

andralex commented:
ok, getting into it

deadalnix commented:
I don't much about DMD source code, so I'll let this to the specialists, but the tests cases sound pretty solid, testing both how the feature is supposed to work and not work.

I'd really like to have that in.

andralex commented on src/aliasthis.c:
```diff
@@ -21,18 +21,371 @@
+ * The the last argument needs for internal use and should be NULL in user call.
+ */
+
+bool iterateAliasThis(Scope *sc, Expression *e, bool (*dg)(Scope *sc, Expression *aliasexpr, void *ctx, Expression **outexpr), void *ctx, Expressions *ret, bool gagerrors, StringTable *directtypes)
```
That's one long-ass line. It can be assumed most folks don't have an editor with good wraparound, so it's better to fix the line lengths.

IgorStepanov replied:
Ok, I'll do it. What a recomended line size? 80 chars?

IgorStepanov replied:
Ok, I've splitted lines, which much larger then 125 chars (The soft limit of origin `expression.c` lines)

yebblies replied:
I suggest pulling out the function pointer type into a well-named typedef, since magicport is going to throw away your nice line-splitting work otherwise.

IgorStepanov replied:
Ok. BTW, when do you plan to move all front-end files to D, without cpp->d porting for each compilation?

IgorStepanov replied:
Fixed.

WalterBright commented:
I'm curious how the following is handled:
```
struct S { void foo(void*); foo(int);}
struct T { void foo(long); }
struct U {
    S s;
    T t;
    alias this s;
    alias this t;
}
U u;
u.foo(null); // calls s.foo(void*)
u.foo(6); // error, s.foo(int) or t.foo(long)?
```
Note that this is how imports work, and the anti-hijacking rules. alias this should work the same way.

IgorStepanov commented:
>I'm curious how the following is handled:

See the [Test18](https://github.com/D-Programming-Language/dmd/pull/3998/files#diff-17b22eae29e74ce6ec29037438b5031cR2194): 
`s.foo(1.0)` and `s.foo("string")` are work correctly, but `s.foo(1);` raises an error, because there are two variants: `s.a.foo(int)` and `s.b.foo(double)`.

If you ask, how it works, there no special case in the algorithm (I'm particularly pleased that it worked as it should before I thought about this case). Algorithm collects all acceptable ways to resolve the expression and if there is more then one way, algorithm raises an error.

deadalnix commented:
That is the most sensible thing to do. I don't have much DMD knowledge, but really would like to use this. I thought it was in already.

IgorStepanov commented:
Please, continue discussion.

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+#include "expression.h"
 
-Expression *resolveAliasThis(Scope *sc, Expression *e)
+typedef Array<const char *> Strings;
```
this is declared in a header somewhere, don't redeclare it

IgorStepanov replied:
Ok, moved to the `arraytypes.h`

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+/**
+ * iterateAliasThis resolves alias this subtypes for `e` and applies it to `dg`.
+ * dg should return true, if appropriate subtype has been found. Otherwise it should returns 0.
+ * Also dg can get context through secong parameter `ctx`, which was passed through fourth argument of iterateAliasThis.
```
these lines are too long for github's display, please word wrap

IgorStepanov replied:
Splitted

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+ * `dg` can return result expression through `outexpr` parameter, and if it is not NULL, it will be pushed to `ret` array.
+ * At the first stage iterateAliasThis iterates direct "alias this"-es and pushes non-null `outexpr`-es to `ret`.
+ * If `dg` for one of direct "alias this"-es returns true then `iterateAliasThis` breaks at this stage and returns true
+ * through return value and returned by `dg` expression array through `ret`.
```
return through `ret`? How does that work?

IgorStepanov replied:
`ret` is a pointer to the `Expression*` array.
User passes it to the `iterateAliasThis` call. `iterateAliasThis` passes to `dg` a ponter to `Expression`. `dg` writes result (if it is found) to this argument, and if `iterateAliasThis` takes non-null expression from `dg`, it pushes it to the `ret` array.
General idea of this function: Only it manages of subtype tree iterating:
now it walks over all subtype tree, however if we will want to change walkup algorithm, we will need only to change this function (not each alias this using).
Current alias this resolve algorithm is not final. 
At least one improvement I can offer now.
We may allow alias this declaration to hide same type alias this declarations of basetypes:

```
struct A
{
    int i;
    alias i this;
}

struct B
{
    int i;
    alias i this;
}

struct C
{
    A a;
    B b;
    int i;
    alias i this;
    alias a this;
    alias b this;
}

void test()
{
    int a = C(); //now this code raises an error: C.i vs C.a.i vs C.b.i
                      //however we may allow to hide C.i of basetype "alias intSymbol this" declarations.
}
```

To make change like this we need only to change `iterateAliasThis` definition, and do not change >9000 alias this related fragments, because each those fragments (for example overloaded operator resolving) doesn't know about subtype tree walkup order. This fragment only defines corresponding delegate which know how to proccess one alias this canditate.

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+ * classes and interfaces.
+ * If one of those `iterateAliasThis` calls returns true our `iterateAliasThis` will return true. Otherwise it will return 0.
+ *
+ * The the last argument needs for internal use and should be NULL in user call.
```
grammar, can't figure out what it means. internal use?

IgorStepanov replied:
I fixed comment. Is it better?

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+    //printf("iterateAliasThis: %s\n", e->toChars());
+
+    Dsymbols *aliasThisSymbols = NULL;
+    if (e->type->toBasetype()->ty == Tstruct)
```
refactor e->type->toBasetype() instead of calling it over and over

IgorStepanov replied:
done

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+    }
+    else
+    {
+        return 0;
```
return false for a bool, not 0

IgorStepanov replied:
done

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+
+    if (!directtypes)
+    {
+        directtypes = new StringTable();
```
the string table contains what?

IgorStepanov replied:
I added corresponding comment.

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+        directtypes = new StringTable();
+        directtypes->_init();
+    }
+    StringValue *depth_counter = directtypes->lookup(e->type->deco, strlen(e->type->deco));
```
make a hash table of types, not of copies of the strings

IgorStepanov replied:
Are you sure that only the one copy of the one type may be here, and type will be merged properly?
I mean that is the `type` pointers equality implies from the `deco` string equality?

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+    return r;
+}
+
+Type *aliasThisOf(Type *t, int idx, bool *islvalue = NULL)
```
please document what this function does

IgorStepanov replied:
done

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+}
+
+/**
+    This function walk over `from` basetype tree and search types,
```
s/This function walk/Walks/

IgorStepanov replied:
done

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+    To prevent infinity loop in types with circular subtyping,
+    pattern "check a flag, lock a flag, do work is a flag wasn't locked, return flag back" is used.
+*/
+MATCH implicitConvToWithAliasThis(Loc loc, Type *from, Type *to, Type *root_from, OutBuffer *buff,
```
many parameters are not documented

IgorStepanov replied:
fixed

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+    if (!buff)
+    {
+        buff = new OutBuffer();
+        buff->writestring(from->toChars());
```
what is buff for and why are strings being written to it?

IgorStepanov replied:
This buff is needed to the properly error messages: 
We need to know name of the currect alias this symbol (`Foo.bar.boo` or `cast(ParentBar)(Foo.bar).moo`.)
And If we have found the one match, we should store the matching symbol name and try to find another matches. 
I'll reflect it in the comment.

IgorStepanov replied:
done

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+    return mret;
+}
+
+void getAliasThisTypes(Type *t, Types *ret, Bools *islvalues)
```
what does this function do?

IgorStepanov replied:
fixed

WalterBright commented on src/aliasthis.c:
```diff
@@ -21,18 +21,376 @@
+    }
+}
+
+Expression *resolveAliasThis(Scope *sc, Expression *e, int num)
```
what does this function do?

IgorStepanov replied:
fixed

WalterBright commented on src/aliasthis.c:
```diff
@@ -69,6 +427,79 @@ Expression *resolveAliasThis(Scope *sc, Expression *e)
 }
 
+/****************************************
+ * Expand alias this tuples.
```
what does this mean?

IgorStepanov replied:
These two functions are taken from current implementation (from [expression.c](https://github.com/D-Programming-Language/dmd/blob/master/src/expression.c#L1017)) and they are used if type contains a tuple alias this.
I do not quite understand this case and I have disallowed multiple alias this if one of the symbols is tuple. Now it works as early.
Let's ask about these functions from their author.

WalterBright commented on src/aliasthis.c:
```diff
@@ -69,6 +427,79 @@ Expression *resolveAliasThis(Scope *sc, Expression *e)
+    return NULL;
+}
+
+int expandAliasThisTuples(Expressions *exps, size_t starti)
```
what does this function do?

IgorStepanov replied:
See the previous comment.

WalterBright commented on src/cast.c:
```diff
@@ -1386,6 +1386,31 @@ Type *toStaticArrayType(SliceExp *e)
 
 /* ==================== castTo ====================== */
+/**
+ * This function should be called by iterateAliasThis.
```
'This function' is implied. 'should be' meaning what happens otherwise?

IgorStepanov replied:
This function is written to be a delegate to `iterateAliasThis`.
C++ doesn't allow us to write function inside another function, thus we can't write delegate declaration "in place".

WalterBright commented on src/cast.c:
```diff
@@ -1386,6 +1386,31 @@ Type *toStaticArrayType(SliceExp *e)
 /* ==================== castTo ====================== */
+/**
+ * This function should be called by iterateAliasThis.
+ * It tries to substitute a next subtype expression and check
```
'tries' meaning what happens if it fails?

IgorStepanov replied:
It will return false and will not write anything to `*outexpr`.
The similar logic can be applied to any `iterateAliasThis` delegates.

WalterBright commented on src/cast.c:
```diff
@@ -2597,7 +2641,180 @@ bool isVoidArrayLiteral(Expression *e, Type *other)
 }
 
+/**
+ * This function should be called by iterateAliasThis.
```
Please remove all 'This function' strings.

IgorStepanov replied:
done. (I hope I did not miss anything)

WalterBright commented on src/cast.c:
```diff
@@ -2597,7 +2641,180 @@ bool isVoidArrayLiteral(Expression *e, Type *other)
+/**
+ * This function should be called by iterateAliasThis.
+ * It is needed to collect of subtypes of the expression type.
+ * It checks if type hasn't been found and returns it if yes.
```
Returns a bool, not 'it'.

IgorStepanov replied:
Comment is fixed.
This function returns false anything, however it also returns expression through `*outexpr` parameter.
Function does it only if  type was found.

WalterBright commented on src/cast.c:
```diff
@@ -2597,7 +2641,180 @@ bool isVoidArrayLiteral(Expression *e, Type *other)
+ * It is needed to collect of subtypes of the expression type.
+ * It checks if type hasn't been found and returns it if yes.
+ */
+static bool atCollectSubtypes(Scope *sc, Expression *e, void *ctx, Expression **outexpr)
```
document what is stored in `*outexpr`

IgorStepanov replied:
Done.

WalterBright commented on src/cast.c:
```diff
@@ -2597,7 +2641,180 @@ bool isVoidArrayLiteral(Expression *e, Type *other)
+    Expressions *candidates = (Expressions *)ctx;
+    bool duplicate = false;
+    assert(e->type);
+    for (int j = 0; j < candidates->dim; ++j)
```
size_t for array indices, not int

IgorStepanov replied:
done

WalterBright commented on src/cast.c:
```diff
@@ -2597,7 +2641,180 @@ bool isVoidArrayLiteral(Expression *e, Type *other)
+    {
+        if ((*candidates)[j]->type->equals(e->type))
+        {
+            duplicate = true;
```
just replace this with `return false;`

IgorStepanov replied:
fixed.

WalterBright commented on src/cast.c:
```diff
@@ -2597,7 +2641,180 @@ bool isVoidArrayLiteral(Expression *e, Type *other)
+    {
+        *outexpr = e;
+    }
+    return false;
```
this function never seems to return true?

IgorStepanov replied:
Yes, it should collect all subtypes, and it shouldn't break the loop by returning true.

WalterBright commented on src/cast.c:
```diff
@@ -2597,7 +2641,180 @@ bool isVoidArrayLiteral(Expression *e, Type *other)
+    return false;
+}
+
+void findAliasThisSubtypes(Scope *sc, Expression *e, Expressions *candidates)
```
what does this function do?

IgorStepanov replied:
comment has been added.

WalterBright commented on src/cast.c:
```diff
@@ -2597,7 +2641,180 @@ bool isVoidArrayLiteral(Expression *e, Type *other)
+
+bool typeMerge(Scope *sc, TOK op, Type **pt, Expression **pe1, Expression **pe2, bool ignorealiastthis = false);
 
+bool mergeAliasThis(Scope *sc, Type **pt, TOK op, Expression **pe1, Expression **pe2)
```
what does this function do?

IgorStepanov replied:
comment has been added.

WalterBright commented on src/cast.c:
```diff
@@ -2597,7 +2641,180 @@ bool isVoidArrayLiteral(Expression *e, Type *other)
+    AggregateDeclaration *ag1 = NULL;
+    AggregateDeclaration *ag2 = NULL;
+
+    if (e1->type->toBasetype()->ty == Tclass)
```
hoist out all the toBasetype() calls

IgorStepanov replied:
fixed

WalterBright commented on src/cast.c:
```diff
@@ -2597,7 +2641,180 @@ bool isVoidArrayLiteral(Expression *e, Type *other)
+    }
+
+    // try to merge e1.aliasthis and e2.aliasthis
+    for (int i = 0; i < candidates1.dim; i++)
```
check all loops for `int` that should be `size_t`

IgorStepanov replied:
done

WalterBright commented on src/cast.c:
```diff
@@ -2609,7 +2826,7 @@ bool isVoidArrayLiteral(Expression *e, Type *other)
  */
 
-bool typeMerge(Scope *sc, TOK op, Type **pt, Expression **pe1, Expression **pe2)
+bool typeMerge(Scope *sc, TOK op, Type **pt, Expression **pe1, Expression **pe2, bool ignorealiastthis)
```
All uses I saw of `ignorealiasthis` are `!ignorealiasthis`. Consider inverting the meaning of the flag to something like `checkaliasthis` or whatever.

WalterBright commented:
This is of terrifying complexity. There's an awful lot of code to implement what should be a rather modest feature. Single alias this is "look it up in alias this, if found, use it". Multiple alias this should be "look it up in each alias this, if it is found exactly once, use it. More than once, issue an error."

andralex commented:
The DIP is good as it stands but 2KLOC spread over 18 files seem quite a hit for what seems to be a small feature.

I think this needs closer review - there are a bunch of comments on this but only a couple about the actual implementation. Without being on the compiler team, I can say that 2KLOC for this can't pass muster.

More reviewers please? Suggestions for simpler angles on the implementation?

IgorStepanov commented:
@andralex @WalterBright 
I had planned to do this tomorrow, but I want to say a few words today.
>Single alias this is "look it up in alias this, if found, use it".

Yes, the old implementation use method "look it up in alias this, if found, use it, otherwise try to look up alias this of alias this".
We may follow this scheme for the multiple alias this: "look it up in each alias this, if it is found - store it, otherwise, recursively apply the next alias this. Eventually, if there is only one alias this stored, use it, if more then one alias this-es are stored, raise an error."
However, I want to encapsulate all walking algorithm over alias this tree, inside the one module: 
if we will change the walking algorithm in future (for example, if we allow to hide subtypes' alias this-es by current type alias this.)
And even if we will create "naive" implementation with dfs, it will not reduce changes size.

>The DIP is good as it stands but 2KLOC spread over 18 files seem quite a hit for what seems to be a small feature.

Unfortunately, we can't much reduce the changes size: simply grep `->aliasthis` in current dmd repository, and horrified. The current alias this implementation spread "over 18 files" and we should modify each of them.

I'll fix my code, add comments over the next few days and I'll able to justify this implementation more substantively.

IgorStepanov commented:
@WalterBright 
>There's a lot of code to implement what should be a rather modest feature. Single alias this is "look it up in alias this, if found, use it". Multiple alias this should be "look it up in each alias this, if it is found exactly once, use it. More than once, issue an error." The implementation here is editing ASTs, has hash tables, I don't see how that follows. What am I missing?

I've partially answer to this early.
Summary:
>The implementation here is editing ASTs

I simplified  AST in this PR. Now each expression has only one flag (aliasthislock) instead of type of the alias this of the next type(s) as early.
Type now contains the aliasthislock bit-set field (as early), but number of cases has beed reduced.

The logic of this PR is simple:
`interateAliasThis` function interates all alias this subtypes of expression.
Only it knows about subtype tree walk logic (it can be easily changed in the future). Hash table need to avoid infinity recursion for circular subtyping case.

Another code use only functions from `aliasthis.c`. When user code uses `interateAliasThis` it should pass delegate to it, and unable to recursive alias this lookup in it. E.g.
```
bool atSubstUna(Scope *sc, Expression *e, void *ctx, Expression **outexpr)
{
    UnaExp *e1 = (UnaExp *)((UnaExp*)ctx)->copy();
    e1->aliasthislock = true; //Now when we do  e1->semantic(), we will not try to resolve alias this.
    e1->e1 = e; //replace op(e1) with op(e1.%aliasthis%)
    Expression *e2 = e1->trySemantic(sc);
    if (e2)
    {
        *outexpr = e2;
        return true;
    }
    return false;
} 
...
                if (!aliasthislock) //we don't need the recursion. Recursion done in iterateAliasThis
                {
                    aliasthislock = true;
                    Expressions results;
                    iterateAliasThis(sc, e1, &atSubstUna, (void*)this, &results);

                    if (results.dim == 1)
                    {
                        return results[0];
                    }
                    else if (results.dim > 1)
                    {
                        error("Unable to unambiguously resolve %s Candidates:", toChars());
                        for (size_t j = 0; j < results.dim; ++j)
                        {
                            error("%s", results[j]->toChars());
                        }
                        return new ErrorExp();
                    }
                }
...
```

I don't allow recursion inside semantic code, because all recursion logic should be incapsulated inside `interateAliasThis`.

@andralex
>The DIP is good as it stands but 2KLOC spread over 18 files seem quite a hit for what seems to be a small feature.
 
This feature is not small:)
It should handle casting, `.member` case, `.method(args)` case including cross-type overloading, `with(symbol)` case, ton cases for overloaded operators (include assignment, casting, foreach, general unary/binary operators), template argument resolving and other cases.
For example, `grep -rnw . -e '>aliasthis' | wc -l` for master returns me 68. And we should handle each those cases.  2K is not much for this task, imho.

IgorStepanov commented:
@WalterBright  @andralex
If you do not go into the fine details, have you fundamental objections aganist this implementation?
I've explained, why I've done so. I have something wrong?

IgorStepanov commented:
ping

deadalnix commented:
Ping ? I need this. Please @andralex and @WalterBright either merge this or propose actionable items to make this move forward. Right now there is not much to do to have this move forward and that is a problem.

IgorStepanov commented:
@andralex AFAIR, you planned to merge this until Jul, 8. Four days remaining:)

ibuclaw commented on src/expression.c:
```diff
@@ -828,6 +946,24 @@ Expression *resolveUFCS(Scope *sc, CallExp *ce)
+                }
+                else if (results.dim > 1)
+                {
+                    ce->error("Unable to unambiguously resolve %s Candidates:", ce->toChars());
```
Errors shouldn't have any capitals in the message, only symbol names have that privilege. ;)

ibuclaw commented on src/expression.c:
```diff
@@ -828,6 +946,24 @@ Expression *resolveUFCS(Scope *sc, CallExp *ce)
+                    ce->error("Unable to unambiguously resolve %s Candidates:", ce->toChars());
+                    for (size_t j = 0; j < results.dim; ++j)
+                    {
+                        ce->error("%s", results[j]->toChars());
```
I'd rather the use of ``errorSupplemental`` here, that way the list of conflicting symbols is correctly indented.

ibuclaw commented:
I agree with @WalterBright and @andralex - there definitely appears to be too much going on, some of which looks like needless noise - such as moving entire functions from ``expression.c`` to ``aliasthis.c`` with no notable benefit - some of which looks like overcomplicating what should be a modest addition.

When looking around at some of the areas that need changing, I was rather taken aback by the spaghetti code used by the existing ``alias this`` implementation.  It must be said that the current design is terrible:

```
Lagain:
    /* 165 lines of code later after checking all other options to resolve the symbol. */
    if (ad && ad->alilasthis)
    {
        e1 = resolveAliasThis(sc, e1);
        goto Lagain;
    }
```
My feeling is that we must first untangle this unwieldy flow of code, only then will it become a "modest addition".

IgorStepanov commented:
@andralex Hello. I was out and will nd I will be absent for some time becauseI'm busy with graduate work at the university and is very busy at the job. I'll return to this work and other D works after a couple of months. Sorry. BTW, I have some ideas about this implementation.

IgorStepanov commented:
@yebblies @andralex 
Hello. Sorry for a long non-availability.
I plan to start revival of this to February.
As I see, D frontend has been moved from cpp to D since the my last activity.
Have you any suggestion, how I may easy move this PR to D using conversion utilite?
What limitation of a frontend code in D?
Can I use foreach, AA or simple dynamic arrays, phobos?

schuetzm commented:
@IgorStepanov Welcome back!

There used to be a script for automatic conversion, but I can't find it anymore. There's also [this page](http://wiki.dlang.org/Updating_Pull_Requests_for_DDMD) on the Wiki about porting PRs to the D version. I guess it's largely no longer relevant now (e.g. magicport.json is gone).

As for which D constructs are allowed: `foreach` is actually preferred over `for`, not sure about AAs and dynamic arrays. Phobos however isn't accepted.

IgorStepanov commented:
Ok. AA may be usefull if it acceptable.
This (C++) version has a magicport staff. I guess, I need to run magicport manually, get .d files and after that, try to merge with current master.
@yebblies can you write command to do magicporting?

yebblies commented:
The procedure in #4922 should work to convert this, once it's rebased onto the last C++ commit using the procedure from the wiki page.

Phobos and the GC are currently being avoided, while some code has been refactored to use foreach. auto and scope delegates.

I would advise doing a direct conversion before any refactoring.

IgorStepanov commented:
Thanks. Looks simpler that I expected. Will do it.

adamdruppe commented:
How is this going?

lesderid commented:
Any update on this?

wilzbach commented:
In the State of D survey, this was among the top five missed language features

![image](https://user-images.githubusercontent.com/4370550/37479264-841c6c46-287c-11e8-8a69-3a763b9569fa.png)

https://rawgit.com/wilzbach/state-of-d/master/report.html

As this is still based on the C++ frontend, it might take a serious effort to revive it.
Though there's some automation - see https://github.com/dlang/dmd/pull/4922#issuecomment-133776696

ibuclaw commented:
This implementation by and large was rejected, as we think that multiple alias this shouldn't be a big change to add support for.

Preliminary PRs may need to be made in order to make the existing alias this support grokable though.  Which is probably why this ended up to unwieldy.

ZombineDev commented:
Closing as #8378 superseded this PR.
